src/main/java/org/apache/gossip/manager/GossipCore.java

```diff
@@ -0,0 +1,243 @@
+        requests.put(t.getUuid() + "/" + t.getUriFrom(), (Base) t);
+      }
+    }
+    if (base instanceof ActiveGossipMessage){
+      List<GossipMember> remoteGossipMembers = new ArrayList<>();
+      RemoteGossipMember senderMember = null;
+      ActiveGossipMessage activeGossipMessage = (ActiveGossipMessage) base;
+      for (int i = 0; i < activeGossipMessage.getMembers().size(); i++) {
+        URI u = null;
+        try {
+          u = new URI(activeGossipMessage.getMembers().get(i).getUri());
+        } catch (URISyntaxException e) {
+          LOGGER.debug("Gossip message with faulty URI", e);
+          continue;
+        }
+        RemoteGossipMember member = new RemoteGossipMember(
+                activeGossipMessage.getMembers().get(i).getCluster(),
+                u,
+                activeGossipMessage.getMembers().get(i).getId(),
+                activeGossipMessage.getMembers().get(i).getHeartbeat());
+       
+        if (i == 0) {
+          senderMember = member;
+        } 
+        if (!(member.getClusterName().equals(gossipManager.getMyself().getClusterName()))){
+          Fault f = new NotAMemberFault("Not a member of this cluster " + i);
+          LOGGER.warn(f);
+          sendOneWay(f, member.getUri());
+          continue;
+        }
+        remoteGossipMembers.add(member);
+      }
+      ActiveGossipOk o = new ActiveGossipOk();
+      sendOneWay(o, senderMember.getUri());
+      mergeLists(gossipManager, senderMember, remoteGossipMembers);
+    }
+  }
+  
+  private void sendInternal(Base message, URI uri){
+    byte[] json_bytes;
+    try {
+      json_bytes = MAPPER.writeValueAsString(message).getBytes();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    int packet_length = json_bytes.length;
+    if (packet_length < GossipManager.MAX_PACKET_SIZE) {
+      byte[] buf = UdpUtil.createBuffer(packet_length, json_bytes);
+      try (DatagramSocket socket = new DatagramSocket()) {
+        InetAddress dest = InetAddress.getByName(uri.getHost());
+        DatagramPacket datagramPacket = new DatagramPacket(buf, buf.length, dest, uri.getPort());
+        socket.send(datagramPacket);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      } 
+    }
+  }
+  
+  public Response send(Base message, URI uri){
+    final Trackable t;
+    if (message instanceof Trackable){
+      t = (Trackable) message;
+      requests.put(t.getUuid(), message);
+    } else {
+      t = null;
+    }
+    sendInternal(message, uri);
+    if (t == null){
+      return null;
+    }
+    final CountDownLatch l = new CountDownLatch(1);
+    final AtomicReference<Response> r = new AtomicReference<>();
+    Thread thread = new Thread(){
+      public void run(){
+        while (true){
+          Base b = requests.get(t.getUuid() + "/" + t.getUriFrom());
+          if (b != null){
+            requests.remove(t.getUuid() + "/" + t.getUriFrom());
+          }
+          r.set((Response) b);
```

Comment:
Good point we do not need a separate get and remove. Actually it is null quite often because if the other host does not respond in 10 seconds null is returned to the caller. We possilbly could catch this and do something else.